Feature/midi enchancement

oleg68 opened the pull request:
Resolves: #703
Resolves: #705
Resolves: #796 

This PR adds new capabilities in the MIDI settings:

1. The choice not to add new midi devices automatically.
2. The choice not to bring the Settings dialog if no active MIDI devices exist at startup.
3. The capability to disable some MIDI APIs.

The enchancements if matching MIDI device names have separated to another issue #885 and will be submited later.

The help page and the screenshot of the ``MIDI devices`` has also been updated.

oleg68 commented:
![midi_devices_tab](https://user-images.githubusercontent.com/11586506/144751058-08bb424d-aa0c-403d-9646-e879c0364480.png)

larspalo commented on help/grandorgue.xml:
```diff
@@ -4929,59 +4914,86 @@ messages to easily switch between samplesets.
+	  <term>MIDI output devices</term>
+	  <listitem>
+	    <para>
+Theis checklistshows all available MIDI output interfaces. Selecting
```
This checklist shows

larspalo commented on src/grandorgue/settings/GOSettingsPorts.h:
```diff
@@ -20,20 +20,53 @@ class GOSettingsPorts
+      m_ApiName = apiName;
+    }
+
+    bool isFor(const wxString& portName, const wxString &apiName = wxEmptyString)
```
isFor is a bit cryptic, I think. If it could it be specified more clearly what this "is for", it would be great.

oleg68 replied:
Renamed to isItemForPortApi

larspalo reviewed (changes requested):
Great job, just some small things you might consider changing.

oleg68 commented:
@larspalo next time you can use ``suggest change`` in the github interface.

oleg68 commented:
The [intermediate build](https://github.com/oleg68/GrandOrgue-official/actions/runs/1541786656) with these changes:

If I open the `File->Settings` dialog and then I press `Cancel` without having made any changes, GO stops listen any MIDI events.

After pressing to the `Panic` button MIDI starts working again.

Seems we need still to work more with this PR.

larspalo commented:
While the latest change fixes the immediate issue, we now have MIDI and sound working in the background when the settings dialog is open, while earlier it was stopped until the settings dialog was closed one way or another.

Now, if the user push the button to change audio device then sound is cut off and continues to be so until settings dialog is closed, even if cancel is selected.

Was there any specific reason why you choose to not stop audio and MIDI when the settings dialog was opened?

oleg68 commented:
> While the latest change fixes the immediate issue, we now have MIDI and sound working in the background when the settings dialog is open, while earlier it was stopped until the settings dialog was closed one way or another.

The same behaviour was before this PR: MIDI and sound was reopened only after closing the setting dialog with `OK` but not with `Cancel`.

oleg68 commented:
Closing MIDI it was not the single issue of this PR before the last change. The most important was the order of MIDI device may be changed that could cause wrong midi map even after `OK` and `Panic`.

Now everything seems ready. @larspalo do you agree with this PR? If it is difficult to revie the whole PR, I could extract changes in sound and moving the device name magic from `GOSoundPortFactory` to `GOPortFactory` to the separate PR.

larspalo reviewed (commented):
I don't mind that the sound and midi is not stopped until really necessary, so that's ok with me.

However, I see that the whitespace handling is somewhat different in some of the files. Like 2 blankspace indentations, brackets placements, and spaces around -> and such. This of course doesn't impact performance, but shouldn't we adhere to the existing project coding standards?

oleg68 commented:
> However, I see that the whitespace handling is somewhat different in some of the files. Like 2 blankspace indentations, brackets placements, and spaces around -> and such. This of course doesn't impact performance, but shouldn't we adhere to the existing project coding standards?

Yes, I use 2-space indenting instead of 8-space (tab) in the totally new code, otherwise it does not fit the line length. But I don't mix different intendation in same functions.

Could you give me examples of wrong bracket placements or wrong spaces around `->`? I'm ready to fix them.

larspalo commented:
GOMidiPortFactory.cpp line 48 et al ( -> )
GOMidiRtPortFactory.cpp line 79 missing whitespace after for (same with 128). Indentations around line 92 are a bit strange (same with 141).
GOPortFactory.cpp line 18 has brackets in java style (like I personally use them...), but most of all it's inconsistent even within the same file.
In GOPortFactory.h you sometimes have brackets like on line 31 and sometimes like on line 34. It's similar in other files.

You'll find similar issues in other files too.

Regarding indentations we earlier did that only by tab in GO code. Personally I've used 2 blankspaces instead as it's more compact, but never in GO code. Now the style of formatting GO code is very diverse...

oleg68 commented:
> GOMidiPortFactory.cpp line 48 et al ( -> ) 

Fixed

> GOMidiRtPortFactory.cpp line 79 missing whitespace after for (same with 128). 

The original code with `for(` was in GOMidiRtInPort.cpp, I've moved it here without adding a space. Now it is fixed.

> Indentations around line 92 are a bit strange (same with 141).

```
	    ports.push_back(
	      new GOMidiRtInPort(
		midi, api, deviceName,
		portFactory.ComposeDeviceName(PORT_NAME, apiName, deviceName)
	      )
	    );
```

The constructor call does not fit into one line. How do you suggest to reformat it?

> GOPortFactory.cpp line 18 has brackets in java style (like I personally use them...), but most of all it's inconsistent even within the same file. 

Fixed.

> In GOPortFactory.h you sometimes have brackets like on line 31 and sometimes like on line 34. It's similar in other files. You'll find similar issues in other files too.

When the statement is short, I put it at the same line. But when the statement is long and does not fit the line, I wrap it to the next line. I prefer not to put a more complex code in the `.h` file. Do you have a better idea of formatting?

> Regarding indentations we earlier did that only by tab in GO code. Personally I've used 2 blankspaces instead as it's more compact, but never in GO code. Now the style of formatting GO code is very diverse...

Not diverse. There are only two: 2 and 8 (tab).

Yes, it does not look right, but now the code is often hard to read because it does not fit the screen. I suggest to change the style to two-spaces. Now I do it only in the new code, but maybe we introduce an automatic clang formatting in the future?

larspalo commented:
As I said earlier, this is not anything essential, but a (reasonably) consistent style makes it easier to read the code.

Personally, if things tend to get too long for a single line I nowadays break it up with one parameter each line:
~~~
whatever someFunctionOrSimilar(
  int firstParameter,
  float secondParameter,
  and so_on,
  until all,
  have been,
  written out
)
{
  ...
}
~~~

But I tend to lean toward java style and would normally have placed the bracket on the same line as the end parenthesis.

I absolutely agree upon not putting anything in the header files that you won't expect to be possible to inline, thus only very short code should be there if at all any.

But, again, to me this is not a really essential topic even if I think it's much more enjoyable to read code that's clear, concise and consistent.

oleg68 commented:
> Personally, if things tend to get too long for a single line I nowadays break it up with one parameter each line:

I do the same in function declarations but not always in function calls.

Should I change

```
	      new GOMidiRtInPort(
		midi, api, deviceName,
		portFactory.ComposeDeviceName(PORT_NAME, apiName, deviceName)
	      )
```

to

```
	      new GOMidiRtInPort(
		midi, 
                api, 
                deviceName,
		portFactory.ComposeDeviceName(PORT_NAME, apiName, deviceName)
	      )
```
?


> But, again, to me this is not a really essential topic even if I think it's much more enjoyable to read code that's clear, concise and consistent.

I'm trying to make the code like this.

Could you reapprove this PR if you do not have essential comments more?

larspalo commented:
> Should I change ... to ... ?

To me, that looks nicer.

oleg68 commented:
> To me, that looks nicer.

OK. I've changed it.

What other changes should I make?

larspalo commented:
Maybe we should let this go for now. There are many other (perhaps much more important) things to focus on, but eventually we should agree upon, and work out (including documenting), a coding standard we intend to follow and over time adapt all project code to.

rousseldenis reviewed (commented):
Seems great!

For code formatting, I use pre-commit and black (for python). But we could use something like https://github.com/pocc/pre-commit-hooks.

So, we don't care about format before committing (after repo configuration).
